### DISCUSSION TITLE
I don't really get what USB CDC means for Tasmota

### DISCUSSION THREAD

**Discussion by hagenbuch** [arendst/Tasmota#19669] [Q&A] [answered]
I am trying to use Tasmota on an ESP32-C3 chip, so I am to choose between tasmota32c3.bin or tasmota32c3cdc.bin

Google says:

"The USB communications device class (CDC) is a composite USB device class, and the class may include more than one interface. The CDC is used primarily for modems, but also for ISDN, fax machines, and telephony applications for performing regular voice calls"

..but I'm sorry I still can't figure out what that means for Tasmota. Which extra devices could that be here? Second network card? And then I read the (precompiled) cdc version has no display capabilities. Does that mean the non-cdc version of tasmota32c3 has them?

Also I would like to know if either of them have the capabilities to interpret modbus code, is USE_SML_M compiled in?

Thanks in advance,..

Andreas

**Comment by barbudor** [comment, marked as answer]
It means that when you connect the USB to the ESP native USB port (instead of through an adapter such as Ch340) it automatically creates a serial port on your computer and the serial console works on it
That remove the need for an adapter on the board and that frees a UART for other purposes

**Reply by hagenbuch** [reply]
Ah thanks!

**Reply by lastmetroid** [reply]
How do you disable the CDC mode on latest tasmota for c3? It seems there are no seperate cdc and non cdc builds anymore. I got a c3 and want to use normal serial tx rx through pin and also want tasmota use these for serial logging. how do I achieve this? I read you need to disable cdc so tasmota uses normal serial rx tx? how?

**Reply by barbudor** [reply]
It is automatic. If you don't connect the USB to a computer, then the console goes to UART0

**Reply by lastmetroid** [reply]
Thank you very much! That information is what I needed. I thought maybe you need to deactivate the usb device first to make it switch.

**Reply by sfromis** [reply]
Restarting a ESP32-C3 board, without connection to a USB host, first message after boot:
> `CMD: Fall back to serial port, no SOF packet detected on USB port`

**Reply by lastmetroid** [reply]
One more question: Is there a way to make a ESP32 C3 work with less than 115200 baudrate for the normal tx/rx pin serial connection, not the usb cdc one? When I set for example Baudrate 56700 it says:

22:27:43.734 CMD: Baudrate 56700
22:27:43.736 SER: Set to 8N1 56700 bit/s
22:27:43.736 SER: HWCDC supports 115200 bit/s only
22:27:43.737 RSL: RESULT = {"Baudrate":56700}

and communication does not work, sending and receiving data to the pins wont work and just garbage comes back. Also a bit confusing why it seems to set it but then at the same time says it wont work.

**Reply by sfromis** [reply]
The above looks a lot like you **are** using USB CDC, where the bit rate is fixed, as the message very clearly says. That is due to CDC not being even similar to UART, but being part of the USB protocol, meaning that the usual concept of signaling speed looses it's meaning.

If you want to use serial for data transfer instead of console, I'd suggest configuring the pins for serial bridge, or even better use the Berry language.

**Reply by lastmetroid** [reply]
No I am not using USB CDC, I just booted the C3 over USB power, not connected to a PC, you said then it would automatically disable the CDC, no? Sending data also works to and from the pins C6 and C7. But just at 115200. Setting lower baudrate wont work and no data is possible anymore to send to the pins.

Not sure what you mean with that using pins for data transfer instead of console. I want to use a Rule like this:

Rule2 ON SerialReceived DO ...

it also works right now, sending a message to the C3 via PIN D7 like this:

![image](https://github.com/arendst/Tasmota/assets/6721083/41629951-0bd8-4dfa-ad7b-f4d2071f4b3e)

But only with Baudrate 115200. Setting lower and it wont work anymore.

As I understood you also can configure custom TX/RX serial pins with software? I tried these commands, they wont work:

23:27:00.395 CMD: SBaudrate 57600
23:27:00.400 RSL: RESULT = {"Command":"Unknown"}
23:27:10.373 CMD: SSerialSend2 test
23:27:10.377 RSL: RESULT = {"Command":"Unknown"}

I need a serial Baudrate of 57600 because of the device I want to connect to the C3 can only work at 57600. The device would just send short strings to the C3, for example:

{"Message":"Test"}

and I wanted to use 2 rules for it like this:

Rule1 ON System#Boot DO Backlog SerialLog 0; SerialSend2 ready ENDON
Rule2 ON SerialReceived#Message DO something with %value% ENDON

What am I missing here? I was not successful to set a lower baudrate, where on another normal ESP32 it works fine. How would I configure 2 other ports (mostly sw serial bridge I guess?) and use a SBaudrate 57600 with the c3?

**Reply by sfromis** [reply]
To try serial bridge, you have to config the pins for serial bridge, `SerBr` instead of `Serial`. And then adapt commands and also rules with the extra `S`.

**Reply by lastmetroid** [reply]
Thank you! That was indeed true, I had to define first SerBr TX and RX to two pins reboot, and now software serial works! I was able to achieve what I wanted now with this:

SBaudrate 57600 (working now)

Rule1 ON System#Boot DO Backlog SerialLog 0; SSerialSend2 ready ENDON
Rule2 ON SSerialReceived#Message DO something with %value% ENDON

Do I need to send a SerialLog 0 every boot too for this to work or is that not necessary? And also a SSerialSend2 at every boot? I guess that initiates the serial bridge to make SSerialReceived work?

**Reply by sfromis** [reply]
No, serial bridge is specifically not the serial console, and not needing to worry about the associated logging option was part of the reason why I suggested changing to serial bridge. I think that you still need to send something before receive is activated, unless you use Berry, which does give you a lot of additional flexibility. 

BTW. It is a bit misleading when the docs suggests this to be "software serial", as **all** serial on ESP32 is hardware serial. It is only on ESP8266 that bit-banging software serial even is a feature of Tasmota. Using the flexibility with just configuring hardware serial where you want makes more sense.

**Reply by lastmetroid** [reply]
So my approach is wrong or not optimal? What disadvantages would I have how I did it? Wouldnt it be stable and reliable? I need it to be 24/7 100% reliable. So how would I do the above with "Berry" I thought rules are Berry already.

**Reply by sfromis** [reply]
I was not trying to suggest anything "wrong", and the word "optimal" has a ton of YMMV. I was just mentioning another way of skinning cats, which can be quite useful in case you want to go beyond a simple rule without much logic, basically depending on your "do something with %value%".

[Berry](https://tasmota.github.io/docs/Berry/) is a programming language, has nothing to do with rules, and provides something comparable to a simplified variant of Python. Of course, everything is a learning curve to get started, but with ESP32, the Berry language is a quite powerful addition to Tasmota over the rather limited "classic" rules.

**Reply by lastmetroid** [reply]
I was looking into the documentation, but unfortunatly the Tasmota documentation is very limited especially for examples it often leaves out important informations. I noticed this now a couple of times, that the documentation is more written from the point of view, that you already know everything already. So you suggest I should writing an entire class / function, which listens to serial ports, then parses the input and does something with it? I dont see how that would be helpful in this case. I would like to get into Berry, but already the first important information is lacking in the Tasmota documentation, how you would add a Berry script as a "listener" which alays runs in the background. It gives a hint, that you could use some autoexec.bat file but it does give no information at all how you do it with examples, how it works at all, or how you would implement a listener script which runs in a loop and scans serial for example all the time but does not waste CPU when it is waiting for serial inputs.

**Reply by sfromis** [reply]
No, I was not making any suggestion, and if the simple rules are ok for you, that's fine.

The general scheme for running Berry with serial is to setup a class with a function invoked at intervals, like every 250 ms, and check if any serial data is available, and read it for processing.

**Comment by sfromis** [comment]
No, scripting is not included in the precompiled binaries for ESP32-C3.

**Reply by hagenbuch** [reply]
Thanks! I guess there is also no unofficial precompiled tasmota32-c3 that has it on so I need to compile myself, right?

**Reply by sfromis** [reply]
No, not in any of the precompiled binaries for ESP32-C3. So, yes.
